File: Services/SafeBrowsingApi/SafeBrowsingApi.cs
using System;
using System.Net.Http;
using System.Text;
using System.Threading.Tasks;
using Newtonsoft.Json;
using shrtlnk.Models.ApiKeys;

namespace shrtlnk.Services.SafeBrowsingApi
{
    public class SafeBrowsingApi
    {
        private readonly HttpClient http;

        private string url = "https://safebrowsing.googleapis.com/v4/threatMatches:find?key=";

        public SafeBrowsingApi(ApiKeys apiKeys, IHttpClientFactory httpFactory)
        {
            http = httpFactory.CreateClient();
            url += apiKeys.SafeBrowsingApi;
        }

        public async Task<bool> CheckUrl(string url)
        {
            Request request = new Request(url);
            Response response = await SendRequest(request);
            return response.matches == null;
        }

        private async Task<Response> SendRequest(Request request)
        {
            try
            {
                string stringifiedPayload = JsonConvert.SerializeObject(request);
                var httpRequest = new HttpRequestMessage
                {
                    Content = new StringContent(stringifiedPayload, Encoding.UTF8, "application/json"),
                    RequestUri = new Uri(url),
                    Method = HttpMethod.Post
                };
                var response = await http.SendAsync(httpRequest);
                return JsonConvert.DeserializeObject<Response>(await response.Content.ReadAsStringAsync());
            }
            catch (Exception e)
            {
                Console.Error.WriteLine("Failed to send request", e);
                return new Response();
            }
        }
    }

    class Request
    {
        public Client client = new Client();
        public ThreatInfo threatInfo;

        public Request(string url)
        {
            threatInfo = new ThreatInfo(url);
        }
    }

    class Client
    {
        public string clientId = "shrtlnk - Josiah Sayers - shrtlnk.dev";
        public string clientVersion = "1.0.0";
    }

    class ThreatInfo
    {
        public string[] threatTypes = new string[]
        {
            "THREAT_TYPE_UNSPECIFIED",
            "MALWARE",
            "SOCIAL_ENGINEERING",
            "UNWANTED_SOFTWARE",
            "POTENTIALLY_HARMFUL_APPLICATION"
        };
        public string[] platformTypes = new string[]
        {
            "ANY_PLATFORM"
        };
        public string[] threatEntryTypes = new string[]
        {
            "THREAT_ENTRY_TYPE_UNSPECIFIED",
            "URL"
        };
        public ThreatEntry[] threatEntries = new ThreatEntry[1];

        public ThreatInfo(string url)
        {
            threatEntries[0] = new ThreatEntry(url);
        }
    }

    class ThreatEntry
    {
        public string url;

        public ThreatEntry(string url)
        {
            this.url = url;
        }
    }

    class Response
    {
        public Array matches { get; set; }
    }
}

File: app/safeBrowsingApi.ts
const apiKey = process.env.SAFE_BROWSING_API_KEY;
if (!apiKey) throw new Error("SAFE_BROWSING_API_KEY not defined");
const apiUrl = `https://safebrowsing.googleapis.com/v4/threatMatches:find?key=${apiKey}`;

export const isUrlSafe = async (url: string): Promise<boolean> => {
  try {
    const request = createRequest(url);
    console.log("Sending request:\n", JSON.stringify(request, null, 2));
    const res = await fetch(apiUrl, {
      body: JSON.stringify(request),
      method: "POST",
    });
    const resData: Response = await res.json();
    console.log("Got response:\n", JSON.stringify(resData, null, 2));
    return (
      !resData.matches ||
      (Array.isArray(resData.matches) && resData.matches.length > 0)
    );
  } catch (e) {
    console.error("Failed to check Safe Browsing API", e);
    return true;
  }
};

function createRequest(url: string) {
  return {
    client: {
      clientId: "shrtlnk - Josiah Sayers - shrtlnk.dev",
      clientVersion: "1.0.0",
    },
    threatInfo: {
      threatTypes: [
        "THREAT_TYPE_UNSPECIFIED",
        "MALWARE",
        "SOCIAL_ENGINEERING",
        "UNWANTED_SOFTWARE",
        "POTENTIALLY_HARMFUL_APPLICATION",
      ],
      platformTypes: ["ANY_PLATFORM"],
      threatEntryTypes: ["THREAT_ENTRY_TYPE_UNSPECIFIED", "URL"],
      threatEntries: { url },
    },
  };
}

type Response = {
  matches: any[];
};
